### DISCUSSION TITLE
CPU of SkyWalking oap is increasing day by day

### DISCUSSION THREAD

**Discussion by peacine** [apache/skywalking#12270] [Q&A]
deployment：docker
oap version：9.5.0
storage type：es 7.10
other config is default

The traffic is stable, but the usage of cpu has grown several times。

Any other infos require？

![image](https://github.com/apache/skywalking/assets/140163342/4bf2c91a-8485-404d-a31f-4b42e01d0dcd)

**Comment by wu-sheng** [comment]
You at least should profile the JVM. This graph doesn't mean anything. Memory could increase due to higher payload, slower of DB performance, etc.

**Reply by peacine** [reply]
Thanks for your quick reply. I find too much CPU occupied by GC in oap. So I modify the JVM config from "-Xmx4g -Xms2g" to "-Xmx6g -Xms6g" and follow up monitoring.

**Reply by peacine** [reply]
> You at least should profile the JVM. This graph doesn't mean anything. Memory could increase due to higher payload, slower of DB performance, etc.

It still appeared. Java options is "-Dmode=no-init -Xmx6g -Xms6g"

**Reply by peacine** [reply]
@wu-sheng @wallezhang @augustowebd @dio Could you help me about this. And I wonder if the partition of `skywalking-segments` kafka topic is 3, is it better to set the number of oap instances to a multiple of 3？

> > You at least should profile the JVM. This graph doesn't mean anything. Memory could increase due to higher payload, slower of DB performance, etc.
> 
> It still appeared. Java options is "-Dmode=no-init -Xmx6g -Xms6g"

**Reply by wu-sheng** [reply]
Personally, I am not a fun of Kafka as transport layer. Poeple added as they preferred. So, don't ask me how to make that works.

Your original asking is OAP having high resource costs, why kafka matters that? confusing.

**Reply by peacine** [reply]
> Personally, I am not a fun of Kafka as transport layer. Poeple added as they preferred. So, don't ask me how to make that works.
> 
> Your original asking is OAP having high resource costs, why kafka matters that? confusing.

Thank you. You are right. I should ask seperately.
The first question is the increasing of cpu cost.
Second is the Kafka and OAP config. As is well known, a partition in Kafka can only be processed by one thread. Is the number of OAP instances related to the number of partitions in Kafka topic？e.g. If partition number of Kafka is 3, the best number of OAP instances is also 3.

**Reply by wu-sheng** [reply]
About Kafka partition, I think so, or, sometimes, more partitions per OAP may be better.
I don't think, as I said, I am not a kafka in the patten.

**Reply by peacine** [reply]
> About Kafka partition, I think so, or, sometimes, more partitions per OAP may be better. I don't think, as I said, I am not a kafka in the patten.

Thanks. How about the first question. I found the frequent gc, but dont know how to solve it. Keeping increasing xmx and xms(6g->8g)? Current Java options is "-Dmode=no-init -Xmx6g -Xms6g"

**Reply by wu-sheng** [reply]
I don't know that. There's not a good memory parameter thing.
It relies on your JVM experiences. 
Maybe bigger is better.
Ask your own dev or platform team Java experts.

**Reply by peacine** [reply]
> I don't know that. There's not a good memory parameter thing. It relies on your JVM experiences. Maybe bigger is better. Ask your own dev or platform team Java experts.

I see.Thank you.
Wish you a good day.

**Reply by wu-sheng** [reply]
About memory, for JVM 17/21, it is common to setup 20-30g memory for high payload, with 12-16 vCPU.

**Reply by peacine** [reply]
> About memory, for JVM 17/21, it is common to setup 20-30g memory for high payload, with 12-16 vCPU.

Thanks very much.
